components/raftstore-v2/src/operation/command/admin/merge/commit.rs

```diff
@@ -0,0 +1,753 @@
+            store_meta
+                .regions
+                .get(&target_region_id)
+                .map(|(r, _)| r.clone())
+        };
+        if let Some(r) = exist_region {
+            let exist_epoch = r.get_region_epoch();
+            let expect_epoch = target_region.get_region_epoch();
+            // exist_epoch > expect_epoch
+            if util::is_epoch_stale(expect_epoch, exist_epoch) {
+                return Err(box_err!(
+                    "target region changed {:?} -> {:?}",
+                    target_region,
+                    r
+                ));
+            }
+            // exist_epoch < expect_epoch
+            if util::is_epoch_stale(exist_epoch, expect_epoch) {
+                info!(
+                    self.logger,
+                    "target region still not catch up, skip.";
+                    "target_region" => ?target_region,
+                    "exist_region" => ?r,
+                );
+                return Ok(false);
+            }
+            return Ok(true);
+        }
+
+        let mut msg = Box::<RaftMessage>::default();
+        msg.set_region_id(target_region.get_id());
+        msg.set_from_peer(self.peer().clone());
+        let local_target_peer = find_peer(target_region, store_ctx.store_id)
+            .cloned()
+            .unwrap();
+        msg.set_to_peer(local_target_peer);
+        msg.set_region_epoch(target_region.get_region_epoch().clone());
+        let extra_msg = msg.mut_extra_msg();
+        extra_msg.set_type(ExtraMessageType::MsgAvailabilityRequest);
+        extra_msg.set_from_region_id(self.region_id());
+        store_ctx
+            .router
+            .force_send_control(StoreMsg::RaftMessage(msg))
+            .unwrap_or_else(|e| {
+                slog_panic!(
+                    self.logger,
+                    "fails to send availability request msg to store";
+                    "error" => ?e,
+                )
+            });
+        Ok(false)
+    }
+
+    pub fn merge_on_availability_response<T>(
+        &mut self,
+        _store_ctx: &mut StoreContext<EK, ER, T>,
+        msg: &RaftMessage,
+    ) {
+        let extra = msg.get_extra_msg();
+        if let Some(state) = self.applied_merge_state()
+            && state.get_target().get_id() == extra.get_from_region_id()
+        {
+            let target_epoch = state.get_target().get_region_epoch();
+            if !extra.has_availability_report()
+                || util::is_epoch_stale(target_epoch, extra.get_availability_report().get_region_epoch())
+            {
+                // TODO: Rollback.
+            }
+            // Else: gather trim status.
+        }
+    }
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    pub fn propose_commit_merge<T>(
+        &mut self,
+        store_ctx: &mut StoreContext<EK, ER, T>,
+        req: RaftCmdRequest,
+    ) -> Result<u64> {
+        if self.storage().has_dirty_data() {
```

Comment:
Actually this is a bit late. Entering prepare merge means unavailable, so dirty check should be pass before prepare merge is proposed.

Reply:
The check will be implemented in a different PR. It will take a few hundred LOC in itself.